internals/interpreter.rst

```diff
@@ -0,0 +1,301 @@
+When a call gets inlined, a new frame gets pushed onto the call stack and the interpreter "jumps" to the start of the callee's bytecode.
+When the callee executes a ``RETURN_VALUE`` instruction, the frame is popped off the call stack and the interpreter returns to the caller.
+There is a flag in the frame (``frame->is_entry``) that indicates whether the frame was inlined (set if it wasn't).
+If ``RETURN_VALUE`` returns to a caller where this flag is set, it performs the usual cleanup and return from ``_PyEval_EvalFrameDefault()``.
+
+A similar check is performed when an unhandled exception occurs.
+
+The call stack
+==============
+
+Up through 3.10 the call stack used to be implemented as a singly-linked list of ``PyFrameObject`` objects.
+This was expensive because each call would require a heap allocation for the stack frame.
+(There was some optimization using a free list, but this was not always effective, because frames are variable length.)
+
+In 3.11 frames are no longer fully-fledged objects.
+Instead, a leaner internal ``_PyInterpreterFrame`` structure is used, which is allocated using a custom allocator, ``_PyThreadState_BumpFramePointer()``.
+Usually a frame allocation is just a pointer bump, which improves memory locality.
+The function ``_PyEvalFramePushAndInit()`` allocates and initializes a frame structure.
+
+Sometimes an actual ``PyFrameObject`` is needed, usually because some Python code calls ``sys._getframe()`` or an extension module calls ``PyEval_GetFrame()``.
+In this case we allocate a proper ``PyFrameObject`` and initialize it from the ``_PyInterpreterFrame``.
+This is a pessimization, but fortunately this happens rarely (introspecting frames is not a common operation).
+
+Things get more complicated when generators are involved, since those don't follow the push/pop model.
+(The same applies to async functions, which are implemented using the same infrastructure.)
+A generator object has space for a ``_PyInterpreterFrame`` structure, including the variable-size part (used for locals and eval stack).
+When a generator (or async) function is first called, a special opcode, ``RETURN_GENERATOR`` is executed, which is responsible for creating the generator object.
+The generator object's ``_PyInterpreterFrame`` is initialized with a copy of the current stack frame.
+The current stack frame is then popped off the stack and the generator object is returned.
+(Details differ depending on the ``is_entry`` flag.)
+When the generator is resumed, the interpreter pushes the ``_PyInterpreterFrame`` onto the stack and resumes execution.
+(There is more hairiness for generators and their ilk, we'll discuss these in a later section in more detail.)
+
+(TODO: Also frame layout and use, and "locals plus".)
+
+All sorts of variables
+======================
+
+The bytecode compiler determines for each variable name in which scope it is defined and generates instructions accordingly.
```

Comment:
```suggestion
The bytecode compiler determines the scope in which each variable name is defined, and generates instructions accordingly.
```

Simplified the original's phrasing and tried to make it much more readible—I had a really hard time parsing it previously.